File: lilac/src/main/java/com/hartwig/hmftools/lilac/sam/SAMSlicer.java
package com.hartwig.hmftools.lilac.sam;

import com.google.common.collect.Lists;
import com.hartwig.hmftools.common.genome.region.GenomeRegion;

import htsjdk.samtools.SAMRecord;
import htsjdk.samtools.SAMRecordIterator;
import htsjdk.samtools.SamReader;

import java.util.List;
import java.util.function.Consumer;
import java.util.stream.Collectors;

public class SAMSlicer
{
    private final int mMinMappingQuality;

    public SAMSlicer(int minMappingQuality)
    {
        mMinMappingQuality = minMappingQuality;
    }

    public final void slice(final GenomeRegion region, final SamReader samReader, final Consumer<SAMRecord> consumer)
    {
        String string = region.chromosome();
        slice(string, (int) region.start(), (int) region.end(), samReader, consumer);
    }

    public final void slice(final String contig, int start, int end, final SamReader samReader, final Consumer<SAMRecord> consumer)
    {
        SAMRecordIterator sAMRecordIterator = samReader.query(contig, start, end, false);

        while(sAMRecordIterator.hasNext())
        {
            SAMRecord samRecord = sAMRecordIterator.next();

            if(!meetsQualityRequirements(samRecord))
                continue;

            consumer.accept(samRecord);
        }
    }

    public final List<SAMRecord> queryMates(final SamReader samReader, final List<SAMRecord> records)
    {
        return records.stream().map(x -> samReader.queryMate(x))
                .filter(x -> x != null)
                .filter(x -> meetsQualityRequirements(x))
                .collect(Collectors.toList());
    }

    private final boolean meetsQualityRequirements(SAMRecord record)
    {
        return record.getMappingQuality() >= mMinMappingQuality
            && !record.getReadUnmappedFlag()
            && !record.getDuplicateReadFlag()
            && !record.isSecondaryOrSupplementary();
    }

}

File: lilac/src/main/kotlin/com/hartwig/hmftools/lilackt/sam/SAMSlicer.kt
package com.hartwig.hmftools.lilackt.sam

import com.hartwig.hmftools.common.genome.region.GenomeRegion
import htsjdk.samtools.SAMRecord
import htsjdk.samtools.SamReader
import java.util.function.Consumer

class SAMSlicer(private val minMappingQuality: Int) {

    fun slice(region: GenomeRegion, samReader: SamReader, consumer: Consumer<SAMRecord>) {
        slice(region.chromosome(), region.start().toInt(), region.end().toInt(), samReader, consumer)
    }

    fun slice(contig: String, start: Int, end: Int, samReader: SamReader, consumer: Consumer<SAMRecord>) {
        val iterator = samReader.query(contig, start, end, false)
        for (samRecord in iterator) {
            if (samRecord.meetsQualityRequirements()) {
                consumer.accept(samRecord)
            }
        }
    }

    fun queryMates(samReader: SamReader, records: List<SAMRecord>): List<SAMRecord> {
        return records
                .mapNotNull { samReader.queryMate(it) }
                .filter { it.meetsQualityRequirements() }
    }

    private fun SAMRecord.meetsQualityRequirements(): Boolean {
        return this.mappingQuality >= minMappingQuality
                && !this.readUnmappedFlag
                && !this.duplicateReadFlag
                && !this.isSecondaryOrSupplementary
    }

}